src/mame/video/snk_bbusters_spr.cpp

```diff
@@ -0,0 +1,236 @@
+// license:BSD-3-Clause
+// copyright-holders:Bryan McPhail
+/***************************************************************************
+
+    Like NeoGeo sprite scale Y line selection is from an external rom.
+
+    For 16 high sprites scale data starts at 0x3800 (9 scale levels)
+    For 32 high sprites scale data starts at 0x7000 (17 scale levels)
+    For 64 high sprites scale data starts at 0xa000 (33 scale levels)
+    For 128 pixel high sprites scale data starts 0xc000 (65 scale levels)
+
+    0xe000 and up - possibly X scale data?  unconfirmed
+
+    Sprites are also double buffered, and this seems to be performed
+    by having two complete sprite chips that are toggled per frame, rather
+    than just ram.  Beast Busters has 4 sprite chips as it has two sprite
+    banks.
+
+***************************************************************************/
+
+#include "emu.h"
+#include "snk_bbusters_spr.h"
+
+DEFINE_DEVICE_TYPE(SNK_BBUSTERS_SPR, snk_bbusters_spr_device, "snk_bbusters_spr", "SNK Beast Busters Sprites")
+
+snk_bbusters_spr_device::snk_bbusters_spr_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: device_t(mconfig, SNK_BBUSTERS_SPR, tag, owner, clock)
+	, device_gfx_interface(mconfig, *this, gfxinfo)
+	, m_scale_table(*this, finder_base::DUMMY_TAG)
+	, m_spriteram(*this, finder_base::DUMMY_TAG)
+{
+}
+
+static const gfx_layout spritelayout =
+{
+	16,16,
+	RGN_FRAC(1,1),
+	4,
+	{ STEP4(0,4) },
+	{
+		STEP4(0,1), STEP4(4*4,1),
+		STEP4(4*4*2*8,1), STEP4(4*4+4*4*2*8,1)
+	},
+	{ STEP8(0,4*4*2), STEP8(16*32,4*8) },
+	128*8
+};
+
+GFXDECODE_MEMBER( snk_bbusters_spr_device::gfxinfo )
+	GFXDECODE_DEVICE( DEVICE_SELF, 0, spritelayout, 0, 16*4 )
+GFXDECODE_END
+
+
+void snk_bbusters_spr_device::device_start()
+{
+	save_item(NAME(m_scale_line_count));
+}
+
+int snk_bbusters_spr_device::adjust4x4(int dx, int dy, int code)
+{
+	if ((dx & 0x10) && (dy & 0x10)) code += 3;
+	else if (dy & 0x10) code += 2;
+	else if (dx & 0x10) code += 1;
+	return code;
+}
+
+int snk_bbusters_spr_device::adjust8x8(int dx, int dy, int code)
+{
+	if ((dx & 0x20) && (dy & 0x20)) code += 12;
+	else if (dy & 0x20) code += 8;
+	else if (dx & 0x20) code += 4;
+	return code;
+}
+
+int snk_bbusters_spr_device::adjust16x16(int dx, int dy, int code)
+{
+	if ((dx & 0x40) && (dy & 0x40)) code += 48;
+	else if (dy & 0x40) code += 32;
+	else if (dx & 0x40) code += 16;
```

Comment:
This should simplify to:
```
	if (dy & 0x40) code += 32;
	if (dx & 0x40) code += 16;
```
16 and 32 sum to 48 anyway.

But I’m seeing a pattern here – if we call the sizes 4×4 = 0, 8×8 = 1, and 16×16 = 2, then all of them simplify down to:
```
	if (dy & (0x10 << Size)) code += 2 << (Size * 2);
	if (dx & (0x10 << Size)) code += 1 << (Size * 2);
```
All three could collapse down to one function with a `Size` template parameter or `size` function parameter.

Reply:
well these were updated from older macros, but yes, it makes more sense to do that